File: Week_02/java/src/_109_ConvertSortedListToBinarySearchTree.java
/**
 * 第一遍：2020/08/19周三 ✅
 * 第二遍：2020/08/19周三
 * 第三遍：2020/06/06周六
 * 第四遍：2020/06/13周六
 */
class _109_ConvertSortedListToBinarySearchTree {
    public TreeNode sortedListToBST(ListNode head) {
        return convert(head, null);
    }

    TreeNode convert(ListNode head, ListNode tail) {
        if (head == tail) {
            return null;
        }
        ListNode slow = head, fast = head;
        while (fast != tail && fast.next != tail) {
            slow = slow.next;
            fast = fast.next.next;
        }

        return new TreeNode(slow.val, convert(head, slow), convert(slow.next, tail));
    }

    class SolutionInorder {
        ListNode node;

        public TreeNode sortedListToBST(ListNode head) {
            node = head;
            int n = 0;
            while (head != null) {
                n++;
                head = head.next;
            }
            return convert(0, n - 1);
        }

        TreeNode convert(int l, int r) {
            if (l > r) return null;
            int m = (l + r + 1) >>> 1;
            TreeNode root = new TreeNode();
            root.left = convert(l, m - 1);
            root.val = node.val;
            node = node.next;
            root.right = convert(m + 1, r);
            return root;
        }
    }

    public class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }

    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: Week_02/kotlin/src/convertsortedarraytobst/_109_ConvertSortedListToBinarySearchTree.kt
package convertsortedarraytobst

fun sortedListToBST(head: ListNode?): TreeNode? {
    fun convert(l: ListNode?, r: ListNode?): TreeNode? {
        if (l == r) return null
        var fast = l
        var slow = l
        while (fast != r && fast?.next != r) {
            fast = fast?.next?.next
            slow = slow?.next
        }

        return TreeNode(slow?.`val`!!, convert(l, slow), convert(slow.next, r))
    }
    return convert(head, null)
}

class _109_Solution2 {
    fun sortedListToBST(head: ListNode?): TreeNode? {
        var len = 0
        var node = head
        while (node != null) {
            node = node.next
            len++
        }
        node = head
        fun convert(l: Int, r: Int): TreeNode? {
            if (l == r) return null
            val m = (l + r) ushr 1
            return TreeNode().apply {
                left = convert(l, m)
                `val` = node?.`val`!!
                node = node?.next
                right = convert(m + 1, r)
            }
        }
        return convert(0, len)
    }
}

class TreeNode(var `val`: Int) {
    constructor(`val`: Int, left: TreeNode?, right: TreeNode?) : this(`val`) {
        this.left = left
        this.right = right
    }

    constructor() : this(0)

    var left: TreeNode? = null
    var right: TreeNode? = null
}

class ListNode(var `val`: Int) {
    var next: ListNode? = null
}